### PR TITLE
Svg Parser: Interactivity

### PR DESCRIPTION
**Pull request by Ian-Stewart-Binks**
This pull request accomplishes the following tasks:
* Issue #342 : The graph is now parsed on the server.
* The SVG output is now nicer.
* SVG graph is now retrieved dynamically.

There are still a few kinks to work out: 
* SVG path curve-to has not been tackled yet. This is the cause of the wonky theory region.
* The graph ID currently has nothing to do with graph retrieval.

### CODE REVIEW

File: hs/ParserUtil.hs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+-- | Gets the root element of the document.
+getRoot :: Document i -> Content i
+getRoot doc = head $ parseDocument (tag "svg") doc
+
+-- | Gets a style attribute from a style String.
+getStyleAttr :: String -> String -> String
+getStyleAttr attr style =
+    drop (length attr + 1) $
+    head $ filter
+            (\x -> take (length attr + 1) x == (attr ++ ":"))
```
Comment (david-yz-liu) [inline hs/ParserUtil.hs]: `startswith`

File: hs/ParserUtil.hs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+-- then this function defaults to the previous style attribute, 'parent'.
+getNewStyleAttr :: String -> String -> String -> String
+getNewStyleAttr newStyle attr parent =
+    if null (getStyleAttr attr newStyle)
+    then parent
+    else getStyleAttr attr newStyle
+
+-- | Converts a tuple of Float to a tuple of Rational.
+convertFloatTupToRationalTup :: (Float, Float) -> (Rational, Rational)
+convertFloatTupToRationalTup tup = (toRational (fst tup), toRational (snd tup))
```
Comment (david-yz-liu) [inline hs/ParserUtil.hs]: pattern matching

File: hs/ParserUtil.hs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+getAttrVal :: Attribute -> String
+getAttrVal ((_,b)) = show b
+
+-- | Converts an Attribute into a more parsable form.
+convertAttributeToTuple :: Attribute -> (String, String)
+convertAttributeToTuple at = (getAttrName at, getAttrVal at)
+
+-- | Gets the children of the current node.
+getChildren :: Content i -> [Content i]
+getChildren = (path [children])
```
Comment (david-yz-liu) [inline hs/ParserUtil.hs]: ()

File: hs/ParserUtil.hs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+processAbsCoords :: [[String]] -> [(Float, Float)]
+processAbsCoords = map convertToFloatTuple
+
+-- | Converts a list of String of length 2 into a tuple of Float.
+convertToFloatTuple :: [String] -> (Float, Float)
+convertToFloatTuple y = (read (head y) :: Float, read (last y) :: Float)
+
+-- | Adds two tuples together.
+addTuples :: (Float, Float) -> (Float, Float) -> (Float, Float)
+addTuples tup1 tup2 = (fst tup1 + fst tup2, snd tup1 + snd tup2)
```
Comment (david-yz-liu) [inline hs/ParserUtil.hs]: pattern matching

File: hs/SVGBuilder.hs [hunk 5]
```diff
@@ -7,46 +7,133 @@ import Control.Monad.IO.Class  (liftIO)
...
-        let pathXml = map (convertPathToXML . buildPath . entityVal) sqlPaths
-        liftIO $ writeFile "Testfile.svg" svgHeader
-        liftIO $ appendFile "Testfile.svg" $ unwords pathXml
-        liftIO $ appendFile "Testfile.svg" $ unwords rectXml
-        liftIO $ appendFile "Testfile.svg" $ unwords textXml
-        liftIO $ appendFile "Testfile.svg" svgFooter
+-- | Builds a Path from a database entry in the paths table.
+buildPaths :: Int -> [Paths] -> [Path]
+buildPaths _ [] = [] 
+buildPaths idCounter entities = do
```
Comment (david-yz-liu) [inline hs/SVGBuilder.hs]: `zipWith`

File: hs/SVGBuilder.hs [hunk 6]
```diff
@@ -7,46 +7,133 @@ import Control.Monad.IO.Class  (liftIO)
...
+    let rectTexts = filter (\x -> intersects
+                            (fromRational (rectsWidth entity))
+                            (fromRational (rectsHeight entity))
+                            (fromRational (rectsXPos entity))
+                            (fromRational (rectsYPos entity))
+                            9
+                            (fromRational (textXPos x))
+                            (fromRational (textYPos x))
+                            ) texts
+    let textString = foldl (++) "" (map textText rectTexts)
```
Comment (david-yz-liu) [inline hs/SVGBuilder.hs]: concat

File: hs/SVGBuilder.hs [hunk 7]
```diff
@@ -7,46 +7,133 @@ import Control.Monad.IO.Class  (liftIO)
...
+         (pathsFill entity)
+         (pathsFillOpacity entity)
+         (pathsStroke entity)
+         (pathsIsRegion entity)
+         ""
+         "" : buildPaths (idCounter + 1) (tail entities)
+
+-- | Builds a Rect from a database entry in the rects table.
+buildRect :: [Text] -> Rects -> Rect
+buildRect texts entity = do
```
Comment (david-yz-liu) [inline hs/SVGBuilder.hs]: Don't use do notation here.

File: hs/SVGBuilder.hs [hunk 8]
```diff
@@ -7,46 +7,133 @@ import Control.Monad.IO.Class  (liftIO)
...
+
+-- | Determines the source and target nodes of the path.
+processPath :: [Rect] -> [Ellipse] -> Path -> Path
+processPath rects ellipses edge = 
+    do let coords = points edge
+       let xStart = fromRational $ fst $ head coords
+       let yStart = fromRational $ snd $ head coords
+       let xEnd = fromRational $ fst $ last coords
+       let yEnd = fromRational $ snd $ last coords
+       let intersectingSourceRect = getIntersectingShape xStart yStart rects
```
Comment (david-yz-liu) [inline hs/SVGBuilder.hs]: Simplify this function.

File: js/graph/parse_graph.js [hunk 9]
```diff
@@ -49,69 +44,7 @@ function buildGraph() {
...
-        });
-
-        if (startNode === '' && endNode === '') {
-            console.log("Could not make edge, diagnose:");
-            console.log("Start: " + startNode);
-            console.log("End: " + endNode);
-        }
-
-        makeEdge(window[startNode], window[endNode], $(this).attr('id'));
+        makeEdge(window[$(this).attr("source-node")], window[$(this).attr("target-node")], $(this).attr('id'));
```
Comment (david-yz-liu) [inline js/graph/parse_graph.js]: ""

File: js/graph/setup.js [hunk 10]
```diff
@@ -81,6 +81,8 @@ var nodes = [];               // List of all nodes
 $(document).ready(function () {
     'use strict';
 
+    $("#graph").append(getRemote());
```
Comment (david-yz-liu) [inline js/graph/setup.js]: append on success

File: hs/ParserUtil.hs (hunk 1 above)
Comment (Ian-Stewart-Binks) [inline hs/ParserUtil.hs]: `startswith` didn't work, using `isPrefixOf`. Done.

File: hs/ParserUtil.hs (hunk 2 above)
Comment (Ian-Stewart-Binks) [inline hs/ParserUtil.hs]: Done.

File: hs/ParserUtil.hs (hunk 3 above)
Comment (Ian-Stewart-Binks) [inline hs/ParserUtil.hs]: Done.

File: hs/ParserUtil.hs (hunk 4 above)
Comment (Ian-Stewart-Binks) [inline hs/ParserUtil.hs]: Done.

File: hs/SVGBuilder.hs (hunk 6 above)
Comment (Ian-Stewart-Binks) [inline hs/SVGBuilder.hs]: Done.

File: hs/GraphResponse.hs [hunk 11]
```diff
@@ -20,8 +20,7 @@ graphResponse =
                  plannerLinks
                 ]
                 (do header "graph"
-                    H.div ! A.id "graph" ! A.class_ "graph" $ do
-                      svgDoc
+                    H.div ! A.id "graph" ! A.class_ "graph" $ do ""
```
Comment (Ian-Stewart-Binks) [inline hs/GraphResponse.hs]: Do.

File: hs/SVGBuilder.hs [hunk 12]
```diff
@@ -54,67 +90,43 @@ printDB = runSqlite dbStr $ do
...
+                                  intersects
+                                  ((ellipsesRx entity) * 2)
+                                  ((ellipsesRy entity) * 2)
+                                  (ellipsesXPos entity, ellipsesYPos entity)
+                                  9
+                                  (textXPos x, textYPos x)
+                                  ) texts
+    Shape ("bool" ++ show idCounter)
+            (ellipsesXPos entity)
+            (ellipsesYPos entity)
```
Comment (Ian-Stewart-Binks) [inline hs/SVGBuilder.hs]: Indentation.

**Comment by Ian-Stewart-Binks** [comment]
@david-yz-liu I think that this is ready for inspection.

**Comment by Ian-Stewart-Binks** [comment]
Mmm. Except that text elements are still using their native positions. Will fix now.

**Pull request merged**